### PR TITLE
ses: Configuration upload & SES cookbook

### PR DESCRIPTION
**Pull request by skazi0**
Added Crowbar UI part for uploading and viewing SES settings based on
yaml file exported from SES cluster.
Settings are saved in a data bag for future use.

This change is a base for: https://github.com/crowbar/crowbar-openstack/pull/2035
and alternative to: https://github.com/crowbar/crowbar-openstack/pull/1977

TODO:
- [ ] validation of uploaded yaml (beyond yaml syntax check). Currently everything is stored in the data bag "as is".
- [ ] CLI interface to upload yaml configs (good for CI)
- [ ] add some service restart logic after the ceph config/keyring files are modified (?)

Some screenshots:
Menu entry:
zrzut ekranu z 2019-02-19 22-46-19

Empty settings view:
zrzut ekranu z 2019-02-19 22-44-28

Upload modal (similar to backup upload):
zrzut ekranu z 2019-02-19 22-44-11

Part of uploaded settings:
zrzut ekranu z 2019-02-19 22-43-50

### CODE REVIEW

**Comment by dirkmueller** [comment]
So far LGTM

File: chef/cookbooks/ses/providers/config.rb [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+# (smart) detection if configs were changed? or is this handled by template somehow?
+
+def write_keyring_file(client_name, keyring_path, key, group_name)
+  Chef::Log.info("SES create #{client_name} keyring #{keyring_path}")
+
+  template keyring_path do
+    cookbook "ses"
+    source "client.keyring.erb"
+    owner "root"
+    group group_name
```
Comment (hemna) [inline chef/cookbooks/ses/providers/config.rb]: The group name should just be ceph, and all of the service users should be added to the ceph group.
Folks can hand edit the yaml file and make all the services use a single client, so you'd end up with glance using the cinder key, so glance needs permissions to read the keyring file.  If they are all part of the same group, then the services can read those keyring files.

After running this the first time and all the services are on the same controller, each keyring ends up with a cinder group permission, which is why glance is failing to create an image.

```
root@d52-54-77-77-01-01:/etc/ceph # ls -al
total 36
drwxr-xr-x   2 root root    4096 Feb 27 14:04 .
drwxr-xr-x 121 root root   12288 Feb 27 14:12 ..
-rw-r-----   1 root cinder    17 Feb 27 14:04 ceph.client..keyring
-rw-r-----   1 root cinder    68 Feb 27 13:52 ceph.client.walt-cinder.keyring
-rw-r-----   1 root cinder    68 Feb 27 13:39 ceph.client.walt-glance.keyring
-rw-r--r--   1 root root     788 Feb 27 14:04 ceph.conf
-rw-r--r--   1 root root      92 Feb 12 17:41 rbdmap

File: chef/cookbooks/ses/providers/config.rb [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+    mode "0755"
+    action :create
+  end
+
+  Chef::Log.info("SES create #{SesHelper.ceph_conf_path}")
+  template SesHelper.ceph_conf_path do
+    cookbook "ses"
+    source "ceph.conf.erb"
+    owner "root"
+    group "root"
```
Comment (hemna) [inline chef/cookbooks/ses/providers/config.rb]: This should be owned by the ceph group too.

Comment (skazi0) [reply chef/cookbooks/ses/providers/config.rb]: done

File: chef/cookbooks/ses/providers/config.rb (hunk 1 above)
Comment (skazi0) [reply chef/cookbooks/ses/providers/config.rb]: should be fixed now

**Comment by jsuchome** [comment]
So the second commit is actually the original @hemna 's ?

File: chef/cookbooks/ses/libraries/helpers.rb [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+      ses_volume_found = false
+      ses_config = ses_settings
+
+      # Loop to check if we have SES managed cluster and update configs
+      cinder_controller[:cinder][:volumes].each_with_index do |volume, volid|
+        next unless volume[:backend_driver] == "rbd" && volume[:rbd][:use_ses]
+
+        # Trying to use_ses but no SES config is available?
+        if ses_config.nil? || !ses_config.key?("cinder")
+          message = "SES configuration was not found but use_ses was enabled!"
```
Comment (jsuchome) [inline chef/cookbooks/ses/libraries/helpers.rb]: this could be more user friendly (I mean some better text instead of "use_ses")

Comment (skazi0) [reply chef/cookbooks/ses/libraries/helpers.rb]: modified

**Comment by skazi0** [comment]
@jsuchome there is no direct "pick" from @hemna's PR. The pieces which fit this alternative design concept were selected by hand. I'm not git-ninja enough to do it using git :o).

**Comment by jsuchome** [comment]
Oh, so this is an alternative? 
Originally I thought it's just UI bits that go together with older PR. 

Now I see you say "This is an alternative version for the Crowbar/Rails side of this functionality". So it's not full replacement, but just the partial one?

**Comment by toabctl** [comment]
I tested the file upload manually and that works as expected. You can query the config options via
```
# knife data bag show crowbar-config ses
default: 
  ses: 
    ceph_conf:     
      cluster_network:      10.84.56.0/21
      fsid:                 d5d7c7cb-5858-3218-a36f-d028df7b0673
      mon_host:             10.84.56.8, 10.84.56.9, 10.84.56.7
      mon_initial_members:  ses-osd1, ses-osd2, ses-osd3
      public_network:       10.84.56.0/21
    cinder:        
      key:             AQCdfIRaxefEMxAAW4zp2My/5HjoST2Y8mJg8A==
      rbd_store_pool:  walt-cinder
      rbd_store_user:  cinder
    cinder-backup: 
      key:             AQBb8hdbrY2bNRAAqJC2ZzR5Q4yrionh7V5PkQ==
      rbd_store_pool:  walt-backups
      rbd_store_user:  cinder-backup
    glance:        
      key:             AQD9eYRachg1NxAAiT6Hw/xYDA1vwSWLItLpgA==
      rbd_store_pool:  walt-glance
      rbd_store_user:  glance
    nova:          
      rbd_store_pool:  walt-nova
    radosgw_urls:  
        http://10.84.56.7:80/swift/v1
        http://10.84.56.8:80/swift/v1
id:       ses
```

**Comment by skazi0** [comment]
@jsuchome the rest is in https://github.com/crowbar/crowbar-openstack/pull/2035 . I added a link to the PR description.

File: chef/cookbooks/ses/resources/config.rb [hunk 4]
```diff
@@ -0,0 +1,22 @@
...
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+actions :create
+
+attribute :service_name, kind_of: String
```
Comment (skazi0) [inline chef/cookbooks/ses/resources/config.rb]: might not be needed at all as we use `new_resource.name`

**Pull request merged**